New-logic

yannl35133 opened the pull request.

lepelog reviewed (changes requested):
haven't looked through everything yet, will continue later

yannl35133 commented on gamepatches.py:
```diff
@@ -368,20 +362,20 @@ class FlagSwitchTypes(IntEnum):
     "Lanayru Gorge",
     "Sealed Grounds",
     "Skyview Temple",
-    "Ancient Cistern",
```
Are those in-game names ? I'd rather make every name the rando uses a constant string, so that no typos can pass typing

lepelog replied:
No, they *mostly* match up with our region names however. So personally I wouldn't tie it to regions and keep it how it is (it doesn't have anything to do with the logic rewrite so do it in a separate PR if you really care about it, these names are potentially only used here)

yannl35133 commented on gamepatches.py:
```diff
@@ -1224,27 +1236,27 @@ def shopsanity_patches(self):
         #     self.startitemflags.append(15)
 
-        if "Progressive Pouch" in self.placement_file.starting_items:
+        if number(PROGRESSIVE_POUCH, 0) in self.placement_file.starting_items:
```
Both this and the starting sword code above rely on this assumption; it will have to be checked by plando

yannl35133 commented on graph_logic/backwards_filled_algorithm.py:
```diff
@@ -0,0 +1,159 @@
+        for exit in unassigned_exits:
+            result = self.logic.relink_connection(exit, entrance, pool)
+            if result:
+                self.link(pool, result, depth + 1)
```
That's ER leftovers, I should just remove it

yannl35133 commented on graph_logic/backwards_filled_algorithm.py:
```diff
@@ -0,0 +1,159 @@
+            return False
+        if not accessible_locations:
+            print(f"No location accessible for {item}")
+            exit(1)
```
There should be a way to properly convey "This seed failed to generate, try another one"

lepelog replied:
Isn't that what I suggested? For example in random settings this could be reached. If you just exit, you can't recover from that

yannl35133 commented on graph_logic/constants.py:
```diff
@@ -0,0 +1,552 @@
+FS_BOSS_KEYS = group(FS_BOSS_KEY, 1)
+SK_BOSS_KEYS = group(SK_BOSS_KEY, 0)
+
+WOODEN_SHIELD = EIN("Wooden Shield")
```
It depends on what you duplicate :
- for reusable items like clawshots, having one is the same as having two, you can just have the claws bit made accessible from 2 locations
- for usable items like keys, it's very different and every single one has to have a bit

lepelog replied:
So the first case would be trivial. Is the second case not possible at all?

yannl35133 commented on graph_logic/constants.py:
```diff
@@ -0,0 +1,552 @@
+FS_MAP = dungeon_map(FS)
+SK_MAP = dungeon_map(SK)
+
+CAVES_KEY = EIN("Lanayru Caves Key")
```
A global search reveals the only place it is used is in `logic/logic.py` which is now dead code.
On a related note, cube sots code uses `"Goddess Cube" in loc` and I would like to remove this to not rely on names at all (besides initially to construct the graph) but I don't know what to replace this with.

lepelog replied:
Cubes have a `cube_region` field in checks.yaml

yannl35133 commented on graph_logic/constants.py:
```diff
@@ -0,0 +1,552 @@
+LMF_NODES_ON_OPTION = EIN("LMF Nodes On option")
+RANDOMIZED_BEEDLE_OPTION = EIN("Randomized Beedle option")
+HERO_MODE = EIN("Hero-mode")
+NO_BIT_CRASHES = EIN("No BiT crashes")
```
Hopefully it will at some point, for now it doesn't hurt much

yannl35133 commented on graph_logic/hints.py:
```diff
@@ -0,0 +1,138 @@
+        hints = self.dist.get_hints(30)
+        self._place_hints_for_locations(hints)
+
+    def _place_hints_for_locations(self, hints: List[GossipStoneHint]):
```
Would be possible, but it would also change how the hints are placed and the topic is too opinionated for me to make (any more) decisions.
Also, adding requirements is not a safe operation, so the whole requirements table would have to be reconstructed from when items were inaccessible (I already kinda do that already though)

yannl35133 commented on graph_logic/hints.py:
```diff
@@ -0,0 +1,138 @@
+        ]
+
+        # ensure prerandomized locations cannot be hinted
+        unhintables = list(self.logic.initial_placement.items.keys())
```
Woops

yannl35133 commented on gamepatches.py:
```diff
@@ -1224,27 +1236,27 @@ def shopsanity_patches(self):
-            self.startstoryflags.append(PROGRESSIVE_SWORD_STORYFLAGS[i])
-        if start_sword_count > 0:
-            self.startitemflags.append(PROGRESSIVE_SWORD_ITEMIDS[start_sword_count - 1])
+        for i in range(NUMBER_SWORDS):
```
Woops

yannl35133 commented on graph_logic/inventory.py:
```diff
@@ -0,0 +1,211 @@
+        v: None
+        | Tuple[int, Set[EXTENDED_ITEM]]
+        | Tuple[str, int]
+        | EXTENDED_ITEM_NAME
```
EIN is the shortened version of EXTENDED_ITEM_NAME when I got tired of how long it is.
If you want to find a better name, it stands for anything that gets assigned a bit in the requirements (so usually items, but actually much more than those)

yannl35133 commented on graph_logic/logic.py:
```diff
@@ -0,0 +1,415 @@
+
+        for loc, req in logic_settings.runtime_requirements.items():
+            it = EXTENDED_ITEM[loc]
+            assert self.opaque[it]
```
I get that it can easily lead to errors if forgotten, but what would you replace it with ?

lepelog replied:
Ideally that would just be a byproduct of organizing the different logic requirements (for example macros can be inlined, exit connections can't, options can be, items can't be) but I don't think that works with the `EXTENDED_ITEM` approach and treating everything equally.

yannl35133 commented on graph_logic/randomize.py:
```diff
@@ -0,0 +1,725 @@
+        useful_checks = [
+            self.placement.items[item]
+            for item in self.get_useful_items(loc, weak)
+            if item not in self.placement.starting_items
```
Swords are numbered, so the ones if your starting inventory are indeed made useless

lepelog commented on checks.yaml:
```diff
@@ -5,87 +5,87 @@ Knight Academy - Fledge's Gift:
   Paths:
     - stage/F009r/r0/l0/TBox/66
-Central Skyloft - Potion Lady's Gift:
+Bazaar - Potion Lady's Gift:
```
Currently, the first part is **always** the region used for hints and the tracker relies on this. This needs to be reverted

lepelog commented on graph_logic/constants.py:
```diff
@@ -0,0 +1,608 @@
+TUMBLEWEED = "Tumbleweed"
+FIVE_BOMBS = "5 Bombs"
+
+BLUE_RUPEES = group(BLUE_RUPEE, 4)
```
It doesn't seem like this list is updated for rupeesanity?

lepelog commented on graph_logic/logic.py:
```diff
@@ -0,0 +1,415 @@
+
+    def _place_item(self, location: EIN, item: EIN, requirements=None):
+        req = DNFInventory(location)
+        if item in self.placement.item_placement_limit and not location.startswith(
+            self.placement.item_placement_limit[item]
+        ):
```
Weren't `EIN` supposed to be treated as opaque strings? But here it's used for dungeon items
Imo it would make more sense to store a set with all the allowed locations of the item in `item_placement_limit` cause that makes it more flexible and plando could use that potentially

lepelog commented on graph_logic/randomize.py:
```diff
@@ -0,0 +1,735 @@
+
+        banned_intset = 0
+        for i in banned_indices:
+            banned_intset += 1 << i
```
```suggestion
            banned_intset |= 1 << i
```
`banned_indices` is typed as `List` so it wouldn't be a misuse of the api to have a duplicate item

lepelog commented on graph_logic/randomize.py:
```diff
@@ -0,0 +1,735 @@
+
+        banned_intset = 0
+        for i in banned_indices:
+            banned_intset += 1 << i
```
```suggestion
            banned_intset |= 1 << i
```

lepelog commented on graph_logic/randomize.py:
```diff
@@ -0,0 +1,735 @@
+        nothing_req = DNFInventory(True)
+        maybe_req = lambda b: banned_req if b else nothing_req
+        self.ban_options = {
+            BEEDLE_STALL_ACCESS: maybe_req(self.options["shop-mode"] == "Always Junk"),
```
Does this not affect Beedle's Crystals as well?

lepelog commented on hints/distributions/Balanced.json:
```diff
@@ -1,19 +1,59 @@
 {
-    "banned_stones": [],
```
Can you revert this autoformatting?

lepelog commented on options.yaml:
```diff
@@ -1,88 +1,106 @@
   permalink: false
   help: Path to folder, where to write the patched wbfs and the spoiler log
+
+- name: No Spoiler Log
+  command: no-spoiler-log
+  type: boolean
+  default: false
+  help: If activated, no spoiler log will be generated. This is highly discouraged
+    to activate, as it makes debugging issues much harder.
+  ui: option_no_spoiler_log
```
Reordering options breaks compatibility between versions (even more than necessary if new options are added), which isn't a problem for rando but it's a big problem for the tracker. Please revert this

lepelog commented on requirements.txt:
```diff
@@ -1,5 +1,7 @@
-PySide6~=6.2.0
\ No newline at end of file
+PySide6~=6.2.0
+lark
```
Can you specify a version here, like the other dependencies do?

lepelog reviewed (changes requested):
When adding stuff to the progress callback, make sure to increase the total accordingly (currently the progress is a bit bugged, it already climbed to 100% quickly and then reset when actually repacking the game) (I hope that's not because of the new-logic-extensions branch, in that case ignore me)

lepelog commented on graph_logic/logic.py:
```diff
@@ -454,19 +474,34 @@ def place_item(self, location: EIN, item: EIN, fill=True):
 
-        self.placement.locations[location] = item
+        if hint_mode:
+            self.placement.stones[location].append(item)
```
Is all of this special casing necessary? Or would it be possible to just create 2 locations per stone and run the normal fill logic on it?

YourAverageLink commented on graph_logic/requirements/Fire Sanctuary.yaml:
```diff
@@ -0,0 +1,242 @@
+    Lower Part:
+      locations:
+        Unlock Key Door: Fire Sanctuary Small Key x 3
+        Move Sliding Door: Mogma Mitts
```
``
```suggestion
        Move Sliding Door: Mogma Mitts & Gust Bellows
```
You need Gust Bellows to blow away the ashes there

yannl35133 commented on checks.yaml:
```diff
@@ -5,92 +5,92 @@ Knight Academy - Fledge's Gift:
-Central Skyloft - Rupee Waterfall Cave Crawlspace:
-  original item: Red Rupee
+Skyloft - Waterfall Cave - Rupee in Crawlspace:
+  original item: "Red Rupee #25"
```
Not really a fan of having the numbers here either, but it makes it possible to generate a partial placement with vanilla items purely out of this file and which will work with any similarly constructed partial placement. I don't know how to do that without embedding the indices

yannl35133 commented:
```diff
diff --git a/CHANGELOG.md b/CHANGELOG.md
index dea5217..4113d07 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,9 +2,9 @@

 ## Dev
 ### Options
+- Added option to make sword reward place swords on heart containers rather than final checks (by YourAverageLink)
 - Added option to randomize progression in Flooded Faron Woods (by CovenEsme)
   - Added Faron SotH check that is obtained when completing the Tadtones story quest
-- Added option to make sword reward place swords on heart containers rather than final checks (by YourAverageLink)
 ### Changes
 - Presets (by cjs07)
   - Presets allow users to save and load their favorite settings quickly from the main page of the randomizer
diff --git a/eventpatches.yaml b/eventpatches.yaml
index 82ec8b7..7131eaf 100644
--- a/eventpatches.yaml
+++ b/eventpatches.yaml
@@ -1005,6 +1005,16 @@
     cases:
       - Rerandomize Treasure
       - -1
+  - name: Skip to end of fledge 2
+    type: flowpatch
+    index: 347 # after giving fledge the second stamina potion
+    flow:
+      next: 353
+  - name: Skip Fledge 3
+    type: flowpatch
+    index: 353
+    flow:
+      next: 320 # before crystal itemget
 116-InsectGame:
   - name: Fix switch 1
     type: flowpatch
@@ -1134,6 +1144,11 @@
       type: type1
       next: Kina Quest Switch
       param4: New Kina Text Option
+  - name: Shorten Kina Text
+    type: flowpatch
+    index: 97
+    flow:
+      next: 103
 118-Town3:
   - name: Shorten Sparrot quest start text
     type: flowpatch
@@ -1184,6 +1199,11 @@
     index: 7
     flow:
       next: 21 # never talk about triforce or song of the hero
+  - name: Shorten Levias text
+    type: flowpatch
+    index: 0
+    flow:
+      next: 22
   - name: Skip bokobase story flag
     type: flowpatch
     index: 22
diff --git a/gui/ui_randogui.py b/gui/ui_randogui.py
index fa0fa7a..7afd751 100644
--- a/gui/ui_randogui.py
+++ b/gui/ui_randogui.py
[Who cares about this file ?]
diff --git a/hints/hint_distribution.py b/hints/hint_distribution.py
index bb9348e..ea38ad1 100644
--- a/hints/hint_distribution.py
+++ b/hints/hint_distribution.py
@@ -97,7 +97,6 @@ class HintDistribution:
         self.barren_dungeons = []
         self.placed_dungeon_barren = 0
         self.prev_barren_type = None
-        self.hinted_locations = []
         self.barren_hinted_areas = set()
         self.counts_by_type = defaultdict(int)

@@ -149,7 +148,7 @@ class HintDistribution:
         self.areas = areas
         self.options = options

-        self.hinted_locations.extend(unhintable)
+        self.hinted_locations = unhintable

         self.banned_stones = list(map(areas.short_to_full, self.banned_stones))
         self.max_hints_per_stone = {
@@ -186,7 +185,7 @@ class HintDistribution:
         self.always_hints = always_hints
         self.sometimes_hints = sometimes_hints

-        # creates a list of boss key locations for required dungeons
+        # creates a list of boss keys for required dungeons
         self.required_boss_keys = [
             boss_key
             for dungeon in self.logic.required_dungeons
@@ -237,6 +236,9 @@ class HintDistribution:
             else:
                 self.barren_overworld_zones.append(zone)

+        self.junk_hints = JUNK_TEXT.copy()
+        self.rng.shuffle(self.junk_hints)
+
         # for each fixed goal hint, place one for each required dungeon
         if "goal" in self.distribution.keys():
             self.distribution["goal"]["fixed"] *= len(self.logic.required_dungeons)
@@ -247,10 +249,7 @@ class HintDistribution:
         for hint_type in self.distribution.keys():
             if self.distribution[hint_type]["fixed"] > 0:
                 needed_fixed.append(hint_type)
-        needed_fixed.sort(key=lambda type: self.distribution[type]["order"])
-
-        self.junk_hints = JUNK_TEXT.copy()
-        self.rng.shuffle(self.junk_hints)
+        needed_fixed.sort(key=lambda hint_type: self.distribution[hint_type]["order"])

         self.hints = []
         for hint_type in needed_fixed:
@@ -346,20 +345,20 @@ class HintDistribution:
             return None

         item = self.hintable_items.pop()
-        location = self.logic.placement.items[item]
+        loc = self.logic.placement.items[item]

-        if location in self.hinted_locations:
+        if loc in self.hinted_locations:
             return self._create_item_hint()
-        self.hinted_locations.append(location)
+        self.hinted_locations.append(loc)

         if self.options["precise-item"]:
-            text = self.areas.checks[location].get("text")
-            return LocationGossipStoneHint("precise_item", location, item, text)
+            text = self.areas.checks[loc].get("text")
+            return LocationGossipStoneHint("precise_item", loc, item, text)

-        if (zone_override := self.areas.checks[location].get("cube_region")) is None:
-            zone_override = self.areas.checks[location]["hint_region"]
+        if (zone_override := self.areas.checks[loc].get("cube_region")) is None:
+            zone_override = self.areas.checks[loc]["hint_region"]

-        return ZoneItemGossipStoneHint(location, item, zone_override)
+        return ZoneItemGossipStoneHint(loc, item, zone_override)

     def _create_random_hint(self):
         all_locations_without_hint = [
@@ -388,6 +387,7 @@ class HintDistribution:
         if not locs:
             if not goal_mode:
                 return None
+
             # if there aren't applicable locations for any goal, return None
             if not any(self.goal_locations):
                 return None
diff --git a/hints/hint_types.py b/hints/hint_types.py
index 97fb317..1dc9796 100644
--- a/hints/hint_types.py
+++ b/hints/hint_types.py
@@ -33,9 +33,6 @@ class GossipStoneHintWrapper:
         secondary_text = self.secondary_hint.to_gossip_stone_text(norm)
         return primary_text + secondary_text

-    def to_spoiler_log_text(self, norm=None) -> str:
-        return f"{self.primary_hint.to_spoiler_log_text(norm)} / {self.secondary_hint.to_spoiler_log_text(norm)}"
-
     def to_spoiler_log_json(self):
         return [
             self.primary_hint.to_spoiler_log_json(),
diff --git a/logic/hints.py b/logic/hints.py
index 0a7adda..12217b9 100644
--- a/logic/hints.py
+++ b/logic/hints.py
@@ -106,7 +106,7 @@ class Hints:
             ]
         )
         self.dist.start(self.logic, needed_always_hints, needed_sometimes_hints)
-        hints = self.dist.get_hints(32)
+        hints = self.dist.get_hints()
         self._place_hints_for_locations(hints)

     def _place_hints_for_locations(self, hints: List[GossipStoneHint]):
diff --git a/logic/logic.py b/logic/logic.py
index 17ceb5e..4bce316 100644
--- a/logic/logic.py
+++ b/logic/logic.py
@@ -108,7 +108,7 @@ class Logic:
                     self.racemode_ban_location(location_name)
                     # print(f'banned {location_name}')

-        if (
+        if self.rando.options["empty-unrequired-dungeons"] and (
             not self.rando.options["triforce-required"]
             or self.rando.options["triforce-shuffle"] == "Anywhere"
         ):
@@ -1382,6 +1382,8 @@ class Logic:
             self.remove_owned_item(item_name)
         for item_name in big_keys_to_place:
             self.remove_owned_item(item_name)
+        for item_name in triforces_to_place:
+            self.remove_owned_item(item_name)

     def place_dungeon_item(self, item_name):
         accessible_undone_locations = self.get_accessible_remaining_locations()
diff --git a/options.yaml b/options.yaml
index 2780037..157b02c 100644
--- a/options.yaml
+++ b/options.yaml
[Look at the diff with pre-new-logic instead]
```
Rebase diff

yannl35133 commented:
[diff.txt](https://github.com/ssrando/ssrando/files/9994270/diff.txt)
Diff for the new commits in Hints Simplification and Chest Dowsing

CovenEsme commented:
Created a PR against the entrance-rando-logic branch to revert the check names. This should solve the 2nd issue blocking new logic from being merged.

CovenEsme commented:
Amended the PR mentioned above to also resolve the 1st issue blocking new logic from being merged.
